homeassistant/components/remootio/cover.py

```diff
@@ -0,0 +1,179 @@
+
+from homeassistant.components import cover
+from homeassistant.components.cover import CoverDeviceClass
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_ENTITY_ID, ATTR_NAME, CONF_DEVICE_CLASS
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import ATTR_SERIAL_NUMBER, CONF_SERIAL_NUMBER, DOMAIN, REMOOTIO_CLIENT
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up an ``RemootioCover`` entity based on the given configuration entry."""
+
+    _LOGGER.debug(
+        "Doing async_setup_entry. config_entry [%s] hass.data[%s][%s] [%s]",
+        config_entry.as_dict(),
+        DOMAIN,
+        config_entry.entry_id,
+        hass.data[DOMAIN][config_entry.entry_id],
+    )
+
+    serial_number: str = config_entry.data[CONF_SERIAL_NUMBER]
+    device_class: CoverDeviceClass = config_entry.data[CONF_DEVICE_CLASS]
+    remootio_client: RemootioClient = hass.data[DOMAIN][config_entry.entry_id][
+        REMOOTIO_CLIENT
+    ]
+
+    async_add_entities(
+        [
+            RemootioCover(
+                serial_number, config_entry.title, device_class, remootio_client
+            )
+        ]
+    )
+
+
+class RemootioCover(cover.CoverEntity):
+    """Cover entity which represents an Remootio device controlled garage door or gate."""
+
+    _remootio_client: RemootioClient
+
+    def __init__(
+        self,
+        unique_id: str,
+        name: str,
+        device_class: CoverDeviceClass,
+        remootio_client: RemootioClient,
+    ) -> None:
+        """Initialize this cover entity."""
+        super().__init__()
+        self._attr_name = name
+        self._attr_unique_id = unique_id
+        self._attr_device_class = device_class
+        self._remootio_client = remootio_client
+        self._attr_device_info = DeviceInfo(
+            name="Remootio",
+            manufacturer="Assemblabs Ltd",
+            suggested_area="The first end-to-end encrypted Wi-Fi and Bluetooth enabled smart remote controller, "
+            "that lets you control and monitor your gates and garage doors using your smartphone.",
+        )
+        self._attr_should_poll = False
+        self._attr_supported_features = cover.SUPPORT_OPEN | cover.SUPPORT_CLOSE
+
+    async def async_added_to_hass(self) -> None:
+        """Register listeners to the used Remootio client to be notified on state changes and events."""
+
+        _LOGGER.debug(
+            "Doing async_added_to_hass for %s with entity id %s",
+            self.__class__,
+            self.entity_id,
+        )
+
+        await self._remootio_client.add_state_change_listener(
+            RemootioCoverStateChangeListener(self)
+        )
```

Comment:
Can these listeners be removed without terminating the client? If so these aught to be removed either with a `@callback` method passed to self.async_on_remove(), or in the async_will_remove_from_hass() function where you terminated the client before.